### PR TITLE
fix(sidecar): don't mutate `account_state` on template refresh

### LINKED ISSUE #505: [CHA-101] Nonce accounting issue
### Describe the bug

There seems to be an issue in the nonce accounting logic in the sidecar.

### Steps to reproduce

* Use `bolt send` to send multiple inclusion requests in a row through the bolt rpc
* Observe error

### Node logs

```text
# Second transaction:

❯ bolt send --private-key $CONTROLLER_KEY --max-fee 2 --priority-fee 2
2024-12-02T11:57:37.333353Z  INFO Target slot: 3103196
2024-12-02T11:57:37.625455Z  INFO tx_hashes=[0x9178f46b141c14f4aa8c4d9271ecaa3ea731519ad0d2602bee985ce5d6119d73] target_slot=3103196 target_sidecar_url=https://rpc-holesky.bolt.chainbound.io/rpc
2024-12-02T11:57:37.824045Z  INFO Response: "{\"jsonrpc\":\"2.0\",\"id\":null,\"error\":{\"code\":-32000,\"message\":\"RPC proxy error: ErrorResponse { jsonrpc: \\\"2.0\\\", id: None, result: Null, error: Some(JsonError { code: Some(Number(-32006)), message: \\\"Transaction nonce too low. Expected 176, got 175\\\", data: None }) }\"}}"

# Node logs:

2024-12-02T12:58:23.560794Z DEBUG POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: Received new request
2024-12-02T12:58:23.560865Z DEBUG POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: New inclusion request inclusion_request=InclusionRequest { slot: 3103502, txs: [FullTransaction { tx: Eip1559 { transaction: TxEip1559 { chain_id: 17000, nonce: 178, gas_limit: 21678, max_fee_per_gas: 2000000000, max_priority_fee_per_gas: 2000000000, to: Call(0xbc58de474b023a40fee115ee097b18987b25841b), value: 100000, access_list: AccessList([]), input: 0x4ae4ccee43b98d321a6532c094d448315b1d33965cd440098481e7016d316ac5 }, signature: PrimitiveSignature { y_parity: true, r: 105226464991429257889957407336659633939565661768785021673835813398731807664079, s: 4718142037388418558295144943397925927559546941978853938859697332232281136131 }, hash: 0x517b4ad09cec7182e2c705edf425b1ad0aa058951406668aa6a3a98a9a31564b }, sender: None }], signature: None, signer: None }
2024-12-02T12:58:23.561487Z  INFO POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: New valid inclusion request received signer=0xbc58de474b023a40fee115ee097b18987b25841b digest=0xfedcc5cc06bc344b126ea407b95214b0ed5d77333c06d876cd6bfb7a5fda4362
2024-12-02T12:58:23.561587Z  INFO bolt_sidecar::driver: Received new commitment request: Inclusion(InclusionRequest { slot: 3103502, txs: [FullTransaction { tx: Eip1559 { transaction: TxEip1559 { chain_id: 17000, nonce: 178, gas_limit: 21678, max_fee_per_gas: 2000000000, max_priority_fee_per_gas: 2000000000, to: Call(0xbc58de474b023a40fee115ee097b18987b25841b), value: 100000, access_list: AccessList([]), input: 0x4ae4ccee43b98d321a6532c094d448315b1d33965cd440098481e7016d316ac5 }, signature: PrimitiveSignature { y_parity: true, r: 105226464991429257889957407336659633939565661768785021673835813398731807664079, s: 4718142037388418558295144943397925927559546941978853938859697332232281136131 }, hash: 0x517b4ad09cec7182e2c705edf425b1ad0aa058951406668aa6a3a98a9a31564b }, sender: None }], signature: Some(Signature { v: NonEip155(false), r: 86990900133513287391998450567076438647132078179886827817022264045076489469138, s: 37530739708679755387145217275469705402076753410699451724691450706253532261364 }), signer: Some(0xbc58de474b023a40fee115ee097b18987b25841b) })
2024-12-02T12:58:23.564220Z DEBUG bolt_sidecar::state::execution: Validating basefee slot_diff=12 basefee=10 max_basefee=56
2024-12-02T12:58:23.564263Z TRACE bolt_sidecar::state::execution: Applying diffs to account state nonce_diff=3 balance_diff=260136000300000
2024-12-02T12:58:23.564280Z DEBUG bolt_sidecar::state::execution: Validating transaction account_state=AccountState { transaction_count: 175, balance: 43703560025847265, has_code: false } nonce_diff=3 balance_diff=260136000300000
2024-12-02T12:58:23.564306Z  INFO bolt_sidecar::driver: Validation against execution state passed target_slot=3103502 elapsed=2.610708ms
2024-12-02T12:58:23.567432Z DEBUG bolt_sidecar::driver: Commitment signed and sent target_slot=3103502 elapsed=5.734994ms
2024-12-02T12:58:23.829150Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:24.001037Z DEBUG bolt_sidecar::state::consensus: Updating slot to 3103492
2024-12-02T12:58:24.638126Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:25.457643Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:26.383469Z  INFO bolt_sidecar::driver: Received new head event slot=3103492
2024-12-02T12:58:26.387855Z TRACE bolt_sidecar::state::execution: Applying execution state update slot=3103492 update=StateUpdate { account_states: {0x20a1305ec6c13ffe8e5289f98ba1a18aa262c15e: AccountState { transaction_count: 1368, balance: 342809207780875002, has_code: false }, 0xbc58de474b023a40fee115ee097b18987b25841b: AccountState { transaction_count: 175, balance: 43703560025847265, has_code: false }}, min_basefee: 10, min_blob_basefee: 1, block_number: 2856649 }
2024-12-02T12:58:26.387906Z TRACE bolt_sidecar::state::execution: Refreshing template... address=0x20A1305Ec6c13FFE8e5289f98ba1A18Aa262c15E account_state=AccountState { transaction_count: 1368, balance: 342809207780875002, has_code: false }
2024-12-02T12:58:26.387926Z TRACE bolt_sidecar::state::execution: Refreshing template... address=0xBC58de474B023a40fEE115Ee097B18987B25841B account_state=AccountState { transaction_count: 175, balance: 43703560025847265, has_code: false }
2024-12-02T12:58:26.390053Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:27.201163Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:28.033259Z DEBUG bolt_sidecar::api::builder: Received register validators request
2024-12-02T12:58:29.289451Z DEBUG POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: Received new request
2024-12-02T12:58:29.289523Z DEBUG POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: New inclusion request inclusion_request=InclusionRequest { slot: 3103502, txs: [FullTransaction { tx: Eip1559 { transaction: TxEip1559 { chain_id: 17000, nonce: 179, gas_limit: 21678, max_fee_per_gas: 2000000000, max_priority_fee_per_gas: 2000000000, to: Call(0xbc58de474b023a40fee115ee097b18987b25841b), value: 100000, access_list: AccessList([]), input: 0x237f753f79b196a190869727ce2fb80f67f570788d1cf661c163c03fe9cda779 }, signature: PrimitiveSignature { y_parity: false, r: 29793288031960769367306473281964470947553967446880316156016426630858080536419, s: 1831567149604977136939996847307580482849295644693424733003195993944646782700 }, hash: 0x39f55ca446b6b9769daaa96bec4a4bcee377f60e02d54613fc18932edde9d982 }, sender: None }], signature: None, signer: None }
2024-12-02T12:58:29.290105Z  INFO POST /rpc{method=bolt_requestInclusion}: bolt_sidecar::api::commitments::handlers: New valid inclusion request received signer=0xbc58de474b023a40fee115ee097b18987b25841b digest=0xcc490bfc9c87b420a406642de05936a900f32b4b2446dbd50694c972cb52646c
2024-12-02T12:58:29.290176Z  INFO bolt_sidecar::driver: Received new commitment request: Inclusion(InclusionRequest { slot: 3103502, txs: [FullTransaction { tx: Eip1559 { transaction: TxEip1559 { chain_id: 17000, nonce: 179, gas_limit: 21678, max_fee_per_gas: 2000000000, max_priority_fee_per_gas: 2000000000, to: Call(0xbc58de474b023a40fee115ee097b18987b25841b), value: 100000, access_list: AccessList([]), input: 0x237f753f79b196a190869727ce2fb80f67f570788d1cf661c163c03fe9cda779 }, signature: PrimitiveSignature { y_parity: false, r: 29793288031960769367306473281964470947553967446880316156016426630858080536419, s: 1831567149604977136939996847307580482849295644693424733003195993944646782700 }, hash: 0x39f55ca446b6b9769daaa96bec4a4bcee377f60e02d54613fc18932edde9d982 }, sender: None }], signature: Some(Signature { v: NonEip155(true), r: 40374869359642513624226197982958173744530324193543258089667883355811196614260, s: 10936873588704194590393359354398523805536805762375062491801577763075606601580 }), signer: Some(0xbc58de474b023a40fee115ee097b18987b25841b) })
2024-12-02T12:58:29.291912Z DEBUG bolt_sidecar::state::execution: Validating basefee slot_diff=10 basefee=10 max_basefee=43
2024-12-02T12:58:29.291947Z TRACE bolt_sidecar::state::execution: Applying diffs to account state nonce_diff=4 balance_diff=346848000400000
2024-12-02T12:58:29.291957Z DEBUG bolt_sidecar::state::execution: Validating transaction account_state=AccountState { transaction_count: 179, balance: 43356712025447265, has_code: false } nonce_diff=4 balance_diff=346848000400000
2024-12-02T12:58:29.291972Z  WARN bolt_sidecar::driver: Execution: failed to validate request err=NonceTooLow(183, 179)
```

### Platform(s)

Linux (x86)

### What version/commit are you on?

v0.3.0-alpha

### Running Mode

Docker mode

### Which chain / network are you on?

Holesky

### Which execution layer client are you using?

Geth

### Which consensus layer client are you using?

Lighthouse

### If you've built bolt from source, provide the full command you used

*No response*

<sub>From [SyncLinear.com](https://synclinear.com) | [CHA-101](https://linear.app/chainbound/issue/CHA-101/nonce-accounting-issue)</sub>

### PR DESCRIPTION
**Pull request by thedevbirb**
Closes #505.

The problem wasn't related to missed slot this time but the logic of refreshing templates.

https://github.com/chainbound/bolt/blob/edaec2964dac70cfcf3bcfad73bd75c62810805d/bolt-sidecar/src/state/execution.rs#L507-L525

It should not `iter_mut` over the `account_states` but rather copying it, which is cheap since it `AccountState` implements `Copy`. 
By mutating it, if two inclusion requests are made for the same `target_slot` but during different slots, then the validation fails because the account state has wrong nonce due to adding block templates nonce and balance diffs to it.

A test to cover this case has been introduced.
Tested on devnet as well

### CODE REVIEW

File: bolt-sidecar/src/state/execution.rs [hunk 1]
```diff
@@ -505,16 +505,16 @@ impl<C: StateFetcher> ExecutionState<C> {
     /// transactions by checking the nonce and balance of the account after applying the state
     /// diffs.
     fn refresh_templates(&mut self) {
-        for (address, (account_state, _)) in self.account_states.iter_mut() {
+        for (address, mut account_state) in self.account_states.iter().map(|(a, (s, _))| (*a, *s)) {
```
Comment (thedevbirb) [inline bolt-sidecar/src/state/execution.rs]: Note: `copied` doesn't make it here because the return type of `iter` is `(&T, &(R, S))` instead of `&T`.

Comment (merklefruit) [reply bolt-sidecar/src/state/execution.rs]: Is there a way to make the behaviour a bit more clear when glancing through? 
It's not _super_ clear that we are mutating the copied version of the state. 

just my 2c, otherwise this is a great catch

File: bolt-sidecar/src/state/execution.rs [hunk 2]
```diff
@@ -1269,4 +1270,67 @@ mod tests {
 
         Ok(())
     }
+
+    /// Sends two inclusion request for the same target slot during two different slots, and makes
+    /// sure the nonces are checked correctly.
+    #[tokio::test]
+    async fn test_subsequent_inclusion_request() -> eyre::Result<()> {
```
Comment (merklefruit) [inline bolt-sidecar/src/state/execution.rs:1283]: this is a fantastic edge case, nice job

File: bolt-sidecar/src/state/execution.rs (hunk 1 above)
Comment (thedevbirb) [reply bolt-sidecar/src/state/execution.rs]: You're right! Added a new commit. Is it clear now in your opinion?

Comment (merklefruit) [reply bolt-sidecar/src/state/execution.rs]: Yes! awesome

**Review by merklefruit — APPROVED** [review]

**Review by mempirate — APPROVED** [review]
Nice catch
